### PR TITLE
add a ClickHouse sink

### PR DESCRIPTION
**Pull request by tommyzli**
```
test-clickhouse-0.test-clickhouse-headless.default.svc.cluster.local :) select * from events \G
...
Row 109:
────────
KubeClusterName:
Reason:                    Started
Message:                   Started container event-exporter
SourceComponent:           kubelet
SourceHost:                kind-worker
FirstTimestamp:            2024-05-15 16:14:16
LastTimestamp:             2024-05-15 16:14:16
Count:                     1
Type:                      Normal
Kind:                      Pod
Namespace:                 default
Name:                      my-release-kubernetes-event-exporter-5747c5b8dd-4lgnx
UID:                       92d13215-23fa-4ae8-a49b-080fff62bac6
ResourceVersion:           251202
FieldPath:                 spec.containers{event-exporter}
Labels:                    {'helm.sh/chart':'kubernetes-event-exporter-3.0.4','pod-template-hash':'5747c5b8dd','app.kubernetes.io/component':'kubernetes-event-exporter','app.kubernetes.io/instance':'my-release','app.kubernetes.io/managed-by':'Helm','app.kubernetes.io/name':'kubernetes-event-exporter','app.kubernetes.io/version':'1.7.0'}
Annotations:               {'checksum/config':'7b57284294d98c65a2318b1d9d462ea267680054ae0e13210c8686912896836d','kubectl.kubernetes.io/restartedAt':'2024-05-15T09:14:15-07:00'}
Deleted:                   false
OwnerReferenceAPIVersions: ['apps/v1']
OwnerReferenceKinds:       ['ReplicaSet']
OwnerReferenceNames:       ['my-release-kubernetes-event-exporter-5747c5b8dd']
OwnerReferenceUIDs:        ['ce97f7c8-a006-4d5e-a3d4-80601c3f2d11']
ReportingComponent:        kubelet
ReportingInstance:         kind-worker
```

### CODE REVIEW

File: pkg/sinks/clickhouse.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	InvolvedObjectAnnotations Map(String, String),
+	InvolvedObjectDeleted Bool,
+	OwnerReferenceAPIVersions Array(LowCardinality(String)),
+	OwnerReferenceKinds Array(LowCardinality(String)),
+	OwnerReferenceNames Array(LowCardinality(String)),
+	OwnerReferenceUIDs Array(UUID),
+	ReportingComponent LowCardinality(Nullable(String)),
+	ReportingInstance Nullable(String),
+) ENGINE = MergeTree
+ORDER BY (LastTimestamp, InvolvedObjectNamespace, InvolvedObjectName)
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go]: Let's reverse this order by (name, namespace, timestamp)

File: pkg/sinks/clickhouse.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+SETTINGS allow_nullable_key=true
+;`
+	insertStatementFmt = `INSERT INTO %s (
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go:62]: let's use async inserts as we're making unbatched inserts

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.20 AS builder
+FROM golang:1.21 AS builder
```
Comment (RoryCrispin) [inline Dockerfile]: If the PR is to be accepted in the upstream repo - we ought not to upgrade the go toolchain

File: go.mod [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
-go 1.20
+go 1.21
+
+toolchain go1.22.2
 
 require (
-	cloud.google.com/go/bigquery v1.44.0
-	cloud.google.com/go/pubsub v1.28.0
+	cloud.google.com/go/bigquery v1.52.0
+	cloud.google.com/go/pubsub v1.32.0
```
Comment (RoryCrispin) [inline go.mod:7]: let's not upgrade the project's dependencies if we plan to contribute back

File: pkg/sinks/clickhouse.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	MaxOpenConns      *int     `yaml:"maxOpenConns"`
+	ConnMaxLifetimeMs *int     `yaml:"connMaxLifetimeMs"`
+	ConnMaxIdleTimeMs *int     `yaml:"connMaxIdleTimeMs"`
+}
+
+const (
+	ddlFmt = `CREATE TABLE %s (
+	KubeClusterName LowCardinality(String),
+	Reason LowCardinality(String),
+	Message String,
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go]: ZSTD

File: pkg/sinks/clickhouse.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	SourceComponent LowCardinality(Nullable(String)),
+	SourceHost Nullable(String),
+	FirstTimestamp Nullable(DateTime),
+	LastTimestamp Nullable(DateTime),
+	Count Nullable(UInt32),
+	Type LowCardinality(Nullable(String)),
+	EventTime Nullable(DateTime),
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go]: Curious why we would want EventTime to be Nullable?

File: pkg/sinks/clickhouse.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+	InvolvedObjectKind LowCardinality(Nullable(String)),
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go]: Do we need InvolvedObeject prefixes? 
If yes, can we have alias columns names IO<thing> (IOKind, IONamespace) to make it a little more ergonomic

File: pkg/sinks/clickhouse.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+	InvolvedObjectLabels Map(String, String),
+	InvolvedObjectAnnotations Map(String, String),
+	InvolvedObjectDeleted Bool,
+	OwnerReferenceAPIVersions Array(LowCardinality(String)),
+	OwnerReferenceKinds Array(LowCardinality(String)),
+	OwnerReferenceNames Array(LowCardinality(String)),
+	OwnerReferenceUIDs Array(UUID),
+	ReportingComponent LowCardinality(Nullable(String)),
+	ReportingInstance Nullable(String),
+) ENGINE = MergeTree
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go]: `PARTITION BY toYYYYMM(EventDate)`

File: pkg/sinks/clickhouse.go (hunk 2 above)
Comment (tommyzli) [reply pkg/sinks/clickhouse.go:62]: I'm setting it in the context: https://github.com/ClickHouse/kubernetes-event-exporter/pull/2/files#diff-47af19dd403d1ce49098e087e5b7f4e64bba34e7283fe095b7ddd5706b6dbd9dR192
Is there a way to do this on the statement level? I got the context variable from our [own examples](https://github.com/ClickHouse/clickhouse-go/blob/v2.24.0/examples/std/async.go)

Comment (RoryCrispin) [reply pkg/sinks/clickhouse.go:62]: missed that, looks good to me!

File: pkg/sinks/clickhouse.go (hunk 6 above)
Comment (tommyzli) [reply pkg/sinks/clickhouse.go]: It's a relatively new field introduced in the v1beta1 api, a bunch of k8s components aren't reporting it yet: https://github.com/kubernetes/kubernetes/issues?q=is%3Aopen+%22use+v1beta1+events%22+is%3Aissue+

File: pkg/sinks/clickhouse.go (hunk 7 above)
Comment (tommyzli) [reply pkg/sinks/clickhouse.go]: I think the IO alias is more confusing to read than it would benefit

File: go.mod (hunk 4 above)
Comment (tommyzli) [reply go.mod:7]: I'm not sure how to prevent go mod tidy from forcing these upgrades

File: pkg/sinks/clickhouse.go (hunk 6 above)
Comment (RoryCrispin) [reply pkg/sinks/clickhouse.go]: Makes sense,
One of the timestamps needs to be non-nullable as we need to associate a time with every event. (What does an event with no time mean? When did it happen?)

At the worst case, we can use the timestamp of the exporter when we receive the event.

File: pkg/sinks/clickhouse.go (hunk 7 above)
Comment (RoryCrispin) [reply pkg/sinks/clickhouse.go]: Sure, can we drop InvolvedObject from the text?
Asking as when you start using this as a power-tool one will wish they didn't need to keep typing InvolvedObjectThis, InvolvedObjectThat 😅

**Review by RoryCrispin — APPROVED** [review]

File: pkg/sinks/clickhouse.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	Annotations Map(String, String),
+	Deleted Bool,
+	OwnerReferenceAPIVersions Array(LowCardinality(String)),
+	OwnerReferenceKinds Array(LowCardinality(String)),
+	OwnerReferenceNames Array(LowCardinality(String)),
+	OwnerReferenceUIDs Array(UUID),
+	ReportingComponent LowCardinality(Nullable(String)),
+	ReportingInstance Nullable(String),
+) ENGINE = %s
+ORDER BY (Namespace, Name, LastTimestamp)
```
Comment (RoryCrispin) [inline pkg/sinks/clickhouse.go]: Name, Namespace, LastTimestamp 

Order by key should be highest cardinality to least (left to right), with Timestamp at the right

**Review by RoryCrispin — APPROVED** [review]

**Pull request merged**